### PR TITLE
Initial version of the shader preloader

### PR DESCRIPTION
**Pull request by gero3**
This is a initial version of a shader precompiler. The shaders can be precompiled with ´renderer.compileMaterials´. And the arguments are scene and camera. I've also added a small example to make check the difference between preloaded shaders and not.
This can be a solution for #10958.

### CODE REVIEW

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1036,6 +1036,47 @@ function WebGLRenderer( parameters ) {
 		return Math.abs( b[ 0 ] ) - Math.abs( a[ 0 ] );
 
 	}
+	
+	this.compileMaterials = function(scene, camera){
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: How about just `compile`?

**Comment by mrdoob** [comment]
Also, would it be possible to do a simpler example? The function itself looks great btw!

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (gero3) [inline src/renderers/WebGLRenderer.js]: This has indeed been changed

**Comment by gero3** [comment]
The example is simplified and I've also added documentation.

**Comment by WestLangley** [comment]
Is it a problem that

    renderer.info.programs.length

produces different results depending on pre-load vs no-pre-load?

**Comment by gero3** [comment]
The only reason there is a difference is because some objects are outside the view. And for those objects the material doesn't get generated directly in without preload. Preload loads all objects in the scene, even if they are invisible on the first frame.

**Comment by brianchirls** [comment]
Looks like this would solve #9887. Thanks for working on it.

**Comment by Astrak** [comment]
Indeed my issue was a duplicate. Thank you for your work on this.

**Comment by gero3** [comment]
Any news on this PR??

**Comment by mrdoob** [comment]
Sorry for the delay 😇

**Comment by mrdoob** [comment]
Thanks!

**Comment by brianchirls** [comment]
Thank you, guys! I really need this.

**Comment by brianchirls** [comment]
A question on this for best practice and possible future improvement:

As I understand it, the reason compiling shaders is janky is that it requires reading back info from the GPU, which causes a delay for sync. If you compile multiple shaders consecutively, will it have to block for multiple frames to wait for each one? Might it make more sense to batch compile all of them and then batch all the status reads together?

**Comment by Astrak** [comment]
Yes no render can happen if the shaders have not all compiled. You can go visit some pages at shadertoy.com that has a lot of big shaders : unless you have one of the best graphic cards, pages will freeze for a while before display. 

I was in need of that compile method to create an [asset wrapper library](https://github.com/Astrak/LoadScreen.js), you can see the compiling step on [the codepen demo](https://github.com/Astrak/LoadScreen.js).

**Pull request merged**